src/crawlee/events/_event_manager.py

```diff
@@ -125,22 +135,41 @@ async def __aexit__(
         await self._emit_persist_state_event_rec_task.stop()
         self._active = False
 
-    def on(self, *, event: Event, listener: Listener) -> None:
+    @overload
+    def on(self, *, event: Literal[Event.PERSIST_STATE], listener: EventListener[EventPersistStateData]) -> None: ...
+    @overload
+    def on(self, *, event: Literal[Event.SYSTEM_INFO], listener: EventListener[EventSystemInfoData]) -> None: ...
+    @overload
+    def on(self, *, event: Literal[Event.MIGRATING], listener: EventListener[EventMigratingData]) -> None: ...
+    @overload
+    def on(self, *, event: Literal[Event.ABORTING], listener: EventListener[EventAbortingData]) -> None: ...
+    @overload
+    def on(self, *, event: Literal[Event.EXIT], listener: EventListener[EventExitData]) -> None: ...
+    @overload
+    def on(self, *, event: Event, listener: EventListener[None]) -> None: ...
+
+    def on(self, *, event: Event, listener: EventListener[Any]) -> None:
```

Comment:
What is the point of keeping types on this non overloaded definition?
In documentation they keep this one without annotations:
https://docs.python.org/3/library/typing.html#typing.overload

If we keep such annotations, then it will probably limit the usefulness of all this overload definitions because this last version is so generic, it will accept almost anything.

Reply:
(btw. Same question for emit method)

Reply:
I am still not sure I understand the use of  EventListener[Any] here. Isn't this too permissive? Shouldn't it be this instead(yes, it is ugly):

EventListener[EventSystemInfoData] | EventListener[EventSystemInfoData] | EventListener[EventMigratingData] | EventListener[EventAbortingData] | EventListener[EventExitData] | EventListener[None]

Reply:
Well, the last method header (the one without @overload) should be a "combination" of all the overloads. Pyright doesn't let you leave out the annotations on it, so I chose to add them. Let me know if that's a problem.

Reply:
Regarding the permissiveness, can you put together an example that shouldn't pass type check but passes?

Reply:
I am just surprised that for example this works for mypy:
```
...
def print_any(whatever: Any) -> None:
    print(whatever)
    
 self._event_manager.on(event=Event.SYSTEM_INFO, listener=print_any)
 ```
 On the other hand I am even more surprised that it works even when you do that explicit list of all options `EventListener[EventSystemInfoData] | EventListener[EventSystemInfoData] | EventListener[EventMigratingData] | EventListener[EventAbortingData] | EventListener[EventExitData] | EventListener[None] `. EventListener[Any] is not included!!!